feat!: Allow reading note hashes and nullifiers from other contracts

sirasistant opened the pull request:
Resolves https://github.com/AztecProtocol/aztec-packages/issues/18084
also resolves https://github.com/AztecProtocol/aztec-packages/issues/18069

LeilaWang commented on noir-projects/aztec-nr/aztec/src/context/private_context.nr:
```diff
@@ -588,6 +590,7 @@ impl PrivateContext {
             min_revertible_side_effect_counter: self.min_revertible_side_effect_counter,
             is_fee_payer: self.is_fee_payer,
             include_by_timestamp: self.include_by_timestamp,
+            // TODO: change this so scope can be configurable
```
Change this to what XD?

sirasistant replied:
old TODO!

LeilaWang commented on noir-projects/noir-contracts/contracts/protocol/contract_instance_registry/src/main.nr:
```diff
@@ -104,8 +103,10 @@ pub contract ContractInstanceRegistry {
         universal_deploy: bool,
     ) {
         // contract class must be published in order to publish an instance
-        self.call(ContractClassRegistry::at(CONTRACT_CLASS_REGISTRY_CONTRACT_ADDRESS)
```
💜

LeilaWang commented on noir-projects/aztec-nr/aztec/src/context/private_context.nr:
```diff
@@ -793,11 +796,30 @@ impl PrivateContext {
+        let note_hash = note_hash_read.note_hash();
+        let contract_address = note_hash_read.contract_address().unwrap_or(AztecAddress::zero());
+
+        self.note_hash_read_requests.any(|r| {
```
I know this is just copying from old code. But maybe we can change it to get a hinted index and return `self.note_hash_read_requests[hinted_index] == note_hash_read`. 
Currently it's only used as `assert(has_been_requested)` by others. And I can't think of an example where an app might do `assert(!has_been_requested)`. So there's no reason to tweak the hinted index and make it return false while there's a match.

sirasistant replied:
I can refactor it to assert_has_been_requested so the interface is not footgunny for users that manually use the context, and implement it with a hinted index!

LeilaWang reviewed (approved):
LGTM!

sirasistant committed:
feat!: Allow reading note hashes and nullifiers from other contracts

Resolves https://github.com/AztecProtocol/aztec-packages/issues/18084
also resolves https://github.com/AztecProtocol/aztec-packages/issues/18069
